### ISSUE TITLE
Visitor + legacyModuleCheck change breaks compat with other transformations

### ISSUE THREAD

**Issue by DvdGiessen** [59naga/babel-plugin-add-module-exports#56]
Since c4aa1b03193d7a8a571e4bc8c93eb52792551bbc this module no longer works correctly with Babel 6.x when plugins such as `babel-plugin-transform-es2015-modules-amd` (and implicitly `babel-plugin-transform-es2015-modules-umd`) are also used (which are quite common to use given this modules use case).

Reason is we're now using the CallExpression visitor, which doesn't get called anymore for the `Object.defineProperty(exports, "__esModule", {value: true});` expression, failing the `isLegacyModule` check. I suspect this is due to these plugins [using a string-based template instead of modifying the AST when building their factory method](https://github.com/babel/babel/blob/6.x/packages/babel-plugin-transform-es2015-modules-amd/src/index.js#L103-L115), and Babel won't reparse that string into an AST, thus denying this plugin the opportunity to hook the CallExpression visitor for it.

Solution may be to use the Program visitor like before, which will ensure we get called at the right time to catch all required information and make our modifications, and ([actually quite similar to how that plugin does it](https://github.com/babel/babel/blob/6.x/packages/babel-plugin-transform-es2015-modules-amd/src/index.js#L81)) use `path.traverse()` to run the legacy module check. (Didn't have time to actually test whether this solution would work.)

**Comment by 59naga** [comment]
Know another way to not alter pure esmodule? it is necessary to check converted that `export default` has been to `exports.default` in `Program.exit`. however, simply writing in the visitor failed.

```js
const result = babelTransform(`export default 'foo'`, {presets: ['env'],
  plugins: [
    (babel) => ({
      visitor: {
        Program: {
          exit (path) {
            const file = path.parent
            const result = generate(file)
            console.log(result.code) // if transformation to commonjs?
          }
        }
      }
    })
  ]})
console.log('===== becomes =====')
console.log(result.code)
```

Within `visitor.Program.exit` can get the code before conversion

```js
"use strict";

export default 'foo';
```
===== becomes =====
```js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = 'foo';
```

Since can't find that way, currently way is search for `Object.defineProperty(exports, "__esModule", {value: true});` with` CallExpression.exit`.

```js
const result = babelTransform(`export default 'foo'`, {presets: ['env'],
  plugins: [
    (babel) => ({
      visitor: {
        CallExpression: {
          exit (path) {
            const file = path.parentPath.parentPath.parent
            const result = generate(file)
            console.log(result.code)
          }
        }
      }
    })
  ]})
console.log('===== becomes =====')
console.log(result.code)
```

```js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = 'foo';
```
===== becomes =====
```js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = 'foo';
```

**Comment by 59naga** [comment]
Can use `path.traverse` if in `Plugin.post`, I'll proceed with the verification.

**Comment by 59naga** [comment]
The hack. add `module.exports` if `AssignmentStatemant` of `exports` only exists `exports.default` when `Plugin.post`. [src/index.js#L30-L40](https://github.com/59naga/babel-plugin-add-module-exports/blob/v0.3.0/src/index.js#L30-L40) published as `v0.3.0` aka `@next`.

this way works with `commonjs` and `umd`, but it doesn't work to `amd`, `systemjs`.
cause they want `module.exports` when `Program.exit`.
I'll wait for Babel to increase the methods for developers. (like [babel/babel#5374](https://github.com/babel/babel/issues/5374))
